### PR TITLE
Make a bunch of classes final and a few private

### PR DESCRIPTION
**Pull request by ceedubs**
Probably best to default to `final` and `private` until we run into a
reason to expose these for extension.

### CODE REVIEW

**Comment by non** [comment]
Looks good to me. :+1:

File: core/src/main/scala/cats/data/Prod.scala [hunk 1]
```diff
@@ -33,35 +33,34 @@ private[data] sealed abstract class ProdInstances extends ProdInstance0 {
...
+private[data] sealed abstract class ProdInstance2 extends ProdInstance3 {
   implicit def prodApplicative[F[_], G[_]](implicit FF: Applicative[F], GG: Applicative[G]): Applicative[Lambda[X => Prod[F, G, X]]] = new ProdApplicative[F, G] {
     def F: Applicative[F] = FF
     def G: Applicative[G] = GG
   }
 }
 
-sealed abstract class ProdInstance3 extends ProdInstance4 {
-  implicit def prodApply[F[_], G[_]](implicit FF: Apply[F], GG: Apply[G]): Apply[Lambda[X => Prod[F, G, X]]] = new ProdApply[F, G] {
+private[data] sealed abstract class ProdInstance3 extends ProdInstance4 { implicit def prodApply[F[_], G[_]](implicit FF: Apply[F], GG: Apply[G]): Apply[Lambda[X => Prod[F, G, X]]] = new ProdApply[F, G] {
```
Comment (travisbrown) [inline core/src/main/scala/cats/data/Prod.scala]: Super nitpicky but do you mind re-inserting the line-break here for the sake of consistency and the diff?

**Comment by travisbrown** [comment]
:+1: from me.

File: core/src/main/scala/cats/data/Prod.scala (hunk 1 above)
Comment (ceedubs) [inline core/src/main/scala/cats/data/Prod.scala]: Good catch. I didn't mean to do this.

Comment (ceedubs) [inline core/src/main/scala/cats/data/Prod.scala]: Hmm weird. This is on two lines for me locally. Am I hitting some bizarre LF/CR/CRLF inconsistency?

**Pull request merged**